libfuzzer/libfuzz_harness.cc

```diff
@@ -42,26 +41,107 @@ int null_printf(FILE* stream, const char* format, ...)
-    std::vector<uint8_t> memory;
+    if (vm == nullptr) {
+        // Failed to create the VM.
+        // This is not interesting, as the fuzzer input is invalid.
+        // Do not add it to the corpus.
+        return vm;
+    }
+
+    ubpf_toggle_undefined_behavior_check(vm.get(), true);
+
+    char* error_message = nullptr;
+
+    ubpf_set_error_print(vm.get(), null_printf);
+
+    if (ubpf_load(vm.get(), program_code.data(), program_code.size(), &error_message) != 0) {
+        // The program failed to load, due to a validation error.
+        // This is not interesting, as the fuzzer input is invalid.
+        // Do not add it to the corpus.
+        free(error_message);
+        vm.reset();
+        return vm;
+    }
+
+    ubpf_toggle_bounds_check(vm.get(), true);
+
+    if (ubpf_register_external_dispatcher(vm.get(), test_helpers_dispatcher, test_helpers_validator) != 0) {
+        // Failed to register the external dispatcher.
+        // This is not interesting, as the fuzzer input is invalid.
+        // Do not add it to the corpus.
+        vm.reset();
+        return vm;
+    }
+
+    if (ubpf_set_instruction_limit(vm.get(), 10000, nullptr) != 0) {
+        // Failed to set the instruction limit.
+        // This is not interesting, as the fuzzer input is invalid.
+        // Do not add it to the corpus.
+        vm.reset();
+        return vm;
+    }
+
+    return vm;
+}
+
+bool call_ubpf_interpreter(const std::vector<uint8_t>& program_code, std::vector<uint8_t>& memory, std::vector<uint8_t>& ubpf_stack, uint64_t& interpreter_result)
+{
+    auto vm = create_ubpf_vm(program_code);
+
+    if (vm == nullptr) {
+        // VM creation failed.
+        return false;
+    }
+
+    // Execute the program using the input memory.
+    if (ubpf_exec_ex(vm.get(), memory.data(), memory.size(), &interpreter_result, ubpf_stack.data(), ubpf_stack.size()) != 0) {
+        // VM execution failed.
+        return false;
+    }
+
+    // VM execution succeeded.
+    return true;
+}
+
+bool call_ubpf_jit(const std::vector<uint8_t>& program_code, std::vector<uint8_t>& memory, std::vector<uint8_t>& ubpf_stack, uint64_t& jit_result)
+{
+    auto vm = create_ubpf_vm(program_code);
+
+    char* error_message = nullptr;
+
+    if (vm == nullptr) {
+        // VM creation failed.
+        return false;
+    }
 
+    auto fn = ubpf_compile_ex(vm.get(), &error_message, JitMode::ExtendedJitMode);
+
+    if (fn == nullptr) {
+        free(error_message);
+
+        // Compilation failed.
+        return false;
+    }
+
+    jit_result = fn(memory.data(), memory.size(), ubpf_stack.data(), ubpf_stack.size());
+
+    // Compilation succeeded.
+    return true;
+}
+
+bool call_linux_jit(const std::vector<uint8_t>& program_code, std::vector<uint8_t>& memory, std::vector<uint8_t> ubpf_stack, uint64_t& linux_jit_result);
+
+bool split_input(const uint8_t* data, std::size_t size, std::vector<uint8_t>& program, std::vector<uint8_t>& memory)
```

Comment:
This function splits the input into object code and then memory's initial contents?

Reply:
Correct.